sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -1909,23 +1909,90 @@ pi_result cuda_piProgramGetInfo(pi_program program, pi_program_info param_name,
   return {};
 }
 
-pi_result cuda_piProgramLink( // TODO: change interface to return error code
+pi_result cuda_piProgramLink(
     pi_context context, pi_uint32 num_devices, const pi_device *device_list,
     const char *options, pi_uint32 num_input_programs,
     const pi_program *input_programs,
     void (*pfn_notify)(pi_program program, void *user_data), void *user_data,
     pi_program *ret_program) {
-  cl::sycl::detail::pi::die("cuda_piProgramLink not implemented");
-  return {};
+  
+  assert(ret_program != nullptr);
+  assert(num_devices == 1 || num_devices == 0);
+  assert(device_list != nullptr || num_devices == 0);
+  assert(pfn_notify == nullptr);
+  assert(user_data == nullptr);
+  pi_result retError = PI_SUCCESS;
+
+  try {
+    ScopedContext active(context);
+
+    CUlinkState state;
+    std::unique_ptr<_pi_program> retProgram{new _pi_program{context}};
+
+    // TODO: Linker options
+    retError = PI_CHECK_ERROR(cuLinkCreate(0, nullptr, nullptr, &state));
+    try {
+      for (size_t i; i < num_input_programs; ++i) {
+        pi_program program = input_programs[i];
+        std::vector<char> source(program->source_,
+                                program->source_ + program->sourceLength_);
+        retError = PI_CHECK_ERROR(cuLinkAddData(
+            state, CU_JIT_INPUT_PTX, source.data(), program->sourceLength_,
+            nullptr, 0, nullptr, nullptr));
+      }
+      void *cubin = nullptr;
+      size_t cubinSize = 0;
+      retError = PI_CHECK_ERROR(cuLinkComplete(state, &cubin, &cubinSize));
+
+      retError = retProgram->create_from_source(
+          static_cast<const char *>(cubin), cubinSize);
+
+      if (retError != PI_SUCCESS) {
+        return retError;
+      }
+
+      retError = retProgram->build_program(options);
+
+      if (retError != PI_SUCCESS) {
+        return retError;
+      }
+    } catch (...) {
+      // Upon error attempt cleanup
+      PI_CHECK_ERROR(cuLinkDestroy(state));
```

Comment:
Correct me If I'm wrong, but PI_CHECK_ERROR can throw one more exception which will lead to abort. Is it expected?

Reply:
You are right in that `PI_CHECK_ERROR` could throw here, but I don't see why this would cause an abort. If that happens it should be caught by the outer try-catch and the error code will be returned by the function.

Arguably, if this happens the error code could be confusing, but then something would be very wrong either way.